Sources/Signal.swift

```diff
@@ -39,131 +45,129 @@ public final class Signal<Value, Error: Swift.Error> {
 					//
 					// Even in the case that both a previous sender and its successor see
-					// the write (the `interruptedState` check before & after the unlock
-					// of `sendLock`), the senders are still bound to the `sendLock`.
-					// So whichever sender loses the battle of acquring `sendLock` is
-					// guaranteed to be blocked.
-					interruptedState = state
-
-					if sendLock.try() {
-						if !terminated, let state = interruptedState {
-							interrupt(state.observers)
+					// `terminating`, the senders are still bound to the `sendLock`. So
+					// whichever sender loses the battle of acquring `sendLock` is
+					// guaranteed to see a status of `terminated`, and eventually be
+					// blocked.
+					signal.status = .terminating(SignalTerminationSnapshot(state: state, event: event))
+
+					if signal.sendLock.try() {
+						if case let .terminating(state) = signal.status {
+							state.send()
+							signal.status = .terminated
 						}
-						sendLock.unlock()
-						signal.generatorDisposable?.dispose()
+						signal.sendLock.unlock()
+						signal.swapDisposable()?.dispose()
 					}
 				}
-			} else {
-				let isTerminating = event.isTerminating
-
-				if let observers = (isTerminating ? signal.state.swap(nil)?.observers : signal.state.value?.observers) {
-					var shouldDispose = false
-
-					sendLock.lock()
+			} else if let observers = signal.state.value?.observers {
+				var shouldDispose = false
 
-					if !terminated {
-						for observer in observers {
-							observer.action(event)
-						}
-
-						// Check if a downstream consumer or a concurrent sender has
-						// interrupted the signal.
-						if !isTerminating, let state = interruptedState {
-							interrupt(state.observers)
-							shouldDispose = true
-						}
+				signal.sendLock.lock()
 
-						if isTerminating {
-							terminated = true
-							shouldDispose = true
-						}
+				if case .alive = signal.status {
+					for observer in observers {
+						observer.action(event)
 					}
 
-					sendLock.unlock()
-
-					// Based on the implicit memory order, any updates to the
-					// `interruptedState` should always be visible after `sendLock` is
-					// released. So we check it again and handle the interruption if
-					// it has not been taken over.
-					if !shouldDispose && !terminated && !isTerminating, let state = interruptedState {
-						sendLock.lock()
-
-						// `terminated` before acquring the lock could be a false negative,
-						// since it might race against other concurrent senders until the
-						// lock acquisition above succeeds. So we have to check again if the
-						// signal is really still alive.
-						if !terminated {
-							interrupt(state.observers)
-							shouldDispose = true
-						}
-
-						sendLock.unlock()
+					// Check if a downstream consumer or a concurrent sender has
+					// interrupted the signal.
+					if case let .terminating(state) = signal.status {
```

Comment:
Do we need this check here with the other check right below?

(If so, I think that'd be helpful to document in the code.)